File: dailyQuestion/2020-05/05-01/python/solution.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


def reverseList(head):
    """
    递归 反转 链表
    :type head: ListNode
    :rtype: ListNode
    """
    if not head :
        return None
    if not head.next :
        return head
    last = reverseList(head.next)
    head.next.next = head
    head.next = None
    return last

def reverseList_items(head) :
    """
    迭代 反转 链表
    :type head: ListNode
    :rtype: ListNode
    """
    pre = None
    current = head
    while current:
        tmp = current.next
        current.next = pre
        pre = current
        current = tmp
    return pre


if __name__ == '__main__':
    node = ListNode(1)
    node.next = ListNode(2)
    node.next.next = ListNode(3)
    node.next.next.next = ListNode(4)
    node.next.next.next.next = ListNode(5)
    result = reverseList(node)
    print()

File: dailyQuestion/2020-05/05-01/golang/solution.go
package main

// Definition for singly-linked list.
type ListNode struct {
    Val int
    Next *ListNode
}


// 用递归反转链表
func reverseList(head *ListNode) *ListNode {
    if head == nil {
        return nil
    }
    if head.Next == nil {
        return head
    }
    last := reverseList(head.Next)
    head.Next.Next = head
    head.Next = nil
    return last
}

// 用迭代的方式反转链表
func reverseList_items(head *ListNode) *ListNode {
    if head == nil {
        return nil
    }
    // 初始化一个空节点
    var pre *ListNode = nil
    current := head
    for current != nil {
        tmp := current.Next
        current.Next = pre
        pre = current
        current = tmp
    }
    return pre
}